### ISSUE TITLE
Ensure only a minimal set of external dependencies vendored for building a hello-world binary

### ISSUE THREAD

**Issue by meteorcloudy** [bazelbuild/bazel#22684]
### Description of the bug:

Using C++ binary as an example.

To build a C++ hello-world binary offline
```
cc_binary(
  name = "bin",
  srcs = ["main.cc"],
)
```
we'll have to vendor the following repositories with Bazel built at 96b90baf08172bb55844c2366111a0c39b6f9e14 or later.

- bazel_features
- bazel_skylib
- platforms
- rules_cc
- rules_java
- rules_python

While, the first four as acceptable, rules_java and rules_python should not be required by default for building a simple C++ hello world program.
``

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

https://github.com/meteorcloudy/minimal_vendor_test/tree/main

### Which operating system are you running Bazel on?

macOS, Linux, Windows

### What is the output of `bazel info release`?

96b90baf08172bb55844c2366111a0c39b6f9e14 or later

**Comment by fmeum** [comment]
Since rules_java and rules_python both define toolchains , I don't see a way to remove them short of getting rid of the dependency from `bazel_tools` all together.

**Comment by meteorcloudy** [comment]
I think we can probably get rid of rules_python in MODULE.tools, I don't see any obvious reason why it should be there.

**Comment by fmeum** [comment]
I think that if we remove it, projects that use `py_binary` without a `load` or dependency on `rules_python` would no longer find a Python toolchain.

**Comment by meteorcloudy** [comment]
Ah, I see, thanks for pointing that out. Then, we can probably only do that in Bazel 8.

**Comment by meteorcloudy** [comment]
@katre @Wyverald As you can see, many dependencies are required only because they register toolchains. Should we consider introducing a way for the root module to exclude toolchains registered by modules in the dependency graph? 

Maybe a simple
```
toolchain_override(
  exclude = "*",
)
```
?

**Comment by katre** [comment]
If the toolchains aren't needed, why is rules_java a dependency? Is it somehow referred to from `@bazel_tools`? Do you have a code link I can check?

**Comment by meteorcloudy** [comment]
yes, rules_java is in MODULE.tools: https://cs.opensource.google/bazel/bazel/+/master:src/MODULE.tools;l=8?q=MODULE.tools&ss=bazel

And the toolchain is registered in rules_java's MODULE.bazel: 
- https://github.com/bazelbuild/rules_java/blob/master/MODULE.bazel#L32C1-L32C20
- https://github.com/bazelbuild/rules_java/blob/master/MODULE.bazel#L90

**Comment by lberki** [comment]
Oof. Do I understand correctly that the root cause is that Python/Java rules require the Python/Java toolchains which are registered in `@rules_python` / `@rules_java`, therefore, if we don't depend on them by default, Python and Java rules cease working. The fix for that is to explicitly require a reference to the rules one needs but that can't happen before Bazel 8 because it would be a very incompatible change?

Owie, then this will require quite a bit of thought to figure out a reasonable (=can be done quickly and doesn't cause too much collateral damage) fix for...

And it is an entirely reasonable request that builds that don't use Python or Java don't depend on their respective rule sets, in fact, it's quite embarrassing that they currently do so...

**Comment by katre** [comment]
It's possible that if the java rules are completely unused, then there is no need for rules_java, it's just very very hard to test that given the current state of things.

Setting up a small toy project and trying to build it with a suitably hacked up bazel is probably the best way to test this.

**Comment by lberki** [comment]
That, I hope, but it doesn't help by itself because there is no way to tell Bazel "I don't need `@rules_java`, pinkie promise, please don't ever look at it", is there?

**Comment by lberki** [comment]
cc @comius -- this discussion might have some bearing on the eventual approach to decoupling rule implementations from the Bazel binary.

**Comment by katre** [comment]
That's essentially what @meteorcloudy was asking for in https://github.com/bazelbuild/bazel/issues/22684#issuecomment-2160601261

> Maybe a simple
> 
> ```
> toolchain_override(
>   exclude = "*",
> )
> ```

I'm very ambivalent about this: I don't think it's a huge amount of work to update [`RegisteredToolchainsFunction`](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/skyframe/toolchains/RegisteredToolchainsFunction.java?ss=bazel&q=RegisteredToolchainsFunction) to handle it, but plumbing the new data through will be tricky (check [the implementation of `register_toolchains` for a comparison](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileGlobals.java?q=symbol%3A%5Cbcom.google.devtools.build.lib.bazel.bzlmod.ModuleFileGlobals.registerToolchains%5Cb%20case%3Ayes)).

But it's very inelegant: it's much better to simply not depend on something you don't intend to use. If `@bazel_tools` current depends on `@rules_java` we're better off trying to break the dependency than just hiding it and papering over.

**Comment by katre** [comment]
Relatedly, are we shipping any source tools as `java_binary` targets?

**Comment by meteorcloudy** [comment]
IIRC, the fundamental problem is that repos defining those toolchains are not fetched lazily because we'll need to fetch the repo to know which toolchain type it is. If this problem is hard to solve in Bazel, then letting users decide which toolchains to be excluded would help avoid unnecessary external dependencies.

**Comment by meteorcloudy** [comment]
> Relatedly, are we shipping any source tools as java_binary targets?

Probably not? /cc @hvadehra

**Comment by fmeum** [comment]
I don't think there is much of a point in artifically cutting the dependency of a trivial C++ hello world on the rules_java repo. When the Java rules have been moved out of Bazel, this dependency will naturally be gone, but any non-trivial Bazel repo will depend on rules_java anyway. 

For example, as soon as you run `bazel coverage` on a trivial C++ example, a Java runtime will be downloaded to run the coverage merger, at which point the actual `rules_java` repo is the least of your worries in terms of vendoring :-) Since ruleset repos are tiny and don't contain binaries, I don't think we need to worry about them too much.

**Comment by Wyverald** [comment]
I agree with Fabian here -- I think the natural solution of "just let Starlarkification run its course" is the best. When all `java_*` rules are loaded from `rules_java`, we can simply remove `rules_java` from `bazel_tools`, which would get rid of the toolchain registration for projects that don't need it. Trying to work around it earlier will likely just add to our tech debt.

Just as an aside: in case this wasn't clear before, we'll likely never be able to completely empty out `bazel_tools`. Stuff like the built-in repo rules (`http_archive` and co.), the shell runfiles library, and the default host platform (`@bazel_tools//tools:host_platform`) will _always_ need to remain bundled with Bazel, even after we've completed Starlarkification.

**Comment by lberki** [comment]
The point of putting in the legwork to make sure that a C++ hello world doesn't depend on Java is that then we know we don't have unexpected dependencies and that it's existence proof that it's possible to run Bazel builds that don't depend on `@rules_java` / `@rules_python`.

I'm not too worried about that one unexpected dependency edge, but I *am* worried about the fact that vendoring `@rules_java` is now pretty much a requirement for builds without Internet access. Firstly, it's a design smell, secondly, I could imagine some trouble brewing if `@rules_java` suddenly decides have a mandatory dependncy on some large Java tool (or a JDK, etc.)

I do realize that `@bazel_tools` will never be completely empty; I wanted to express myself concisely instead of precisely. I'm sure we all agree, though, that right now it has _way_ more things in it than it should have.

**Comment by fmeum** [comment]
We do already have tests that verify that `cc_test` doesn't depend on certain unexpected tools: https://cs.opensource.google/bazel/bazel/+/0dbfaccaf5bee5ea7f11c01db1fc0cd1ca7f3810:src/test/shell/bazel/cc_integration_test.sh;l=1683.
We could extend this to ensure that there are no unexpected rules_java deps without getting rid of the toolchain registration edge. 

Doing that via a `bazel vendor` test would be both more effort and less granular (repo vs target level). If we also add a test to rules_java that detects undesired edges to other repos, we could even catch any accidentally eager fetches of toolchains early.

**Comment by meteorcloudy** [comment]
I think the problem we want to solve is this: **For users who want to vendor external dependencies to achieve offline build, they don't have to vendor repos they don't need just because of toolchain registrations.**

This is important for projects like [Kleaf](https://android.googlesource.com/kernel/build/+/refs/heads/main/kleaf/README.md) which has a strict policy on vendoring 3rd party dependency.

And it's still a problem after we starlarkifying rules_java and move it out of bazel_tools. For example, if kleaf depends on rules_foo, and rules_foo depends on rules_java because a binary tool needs it to provide some optional feature. However, Kleaf knows it doesn't want the optional feature and doesn't need Java toolchains anywhere else. Currently, there is no nice way to avoid vendoring rules_java and potentially other repos introudced during toolchain registration (e.g. [bazel_features](https://github.com/bazelbuild/rules_java/commit/7520a693c4af635e8f485c6a70c15f03bb929011)).

I think the solution might be to offer a way for the root module to exclude toolchain registrations from given modules. Something like
```
toolchain_exclude(
   module_names = ["rules_java"],
   # maybe also: exclude_all = True,
)
```

> We do already have tests that verify that cc_test doesn't depend on certain unexpected tools: https://cs.opensource.google/bazel/bazel/+/0dbfaccaf5bee5ea7f11c01db1fc0cd1ca7f3810:src/test/shell/bazel/cc_integration_test.sh;l=1683.

This test doesn't guard what repos need to be fetched for building the target. For two reasons: 1) toolchain registration 2) irrelevant load statements. That's why we had to traverse the skyframe graph to discover all repos to vendor for given targets. Therefore, I think `bazel vendor` is the only way to discover full repo dependencies for targets so far.

**Comment by meteorcloudy** [comment]
> we'll likely never be able to completely empty out bazel_tools

I think the actual goal is to just make MODULE.tools empty instead of making the whole bazel_tools empty? So far, I only see `bazel_dep(name = "buildozer", version = "7.1.2")` to be a problem, but it has little negative impact.

**Comment by lberki** [comment]
Yeah, I'm supportive of the idea of being able to say "to the best of knowledge, my builds don't need Java and please error out if it touches anything that touches Java".

Practically speaking, I could imagine a knob in MODULE.bazel that says "depend on these builtin language families" and if the user doesn't say anything, they get all of them. This would provide a convenient way to cut off rule implementations from Bazel if we eventually switch the default to "nothing".

I haven't thought the mechanics through, though. Maybe @comius knows with certainty whether this is a good idea?

**Comment by fmeum** [comment]
> Yeah, I'm supportive of the idea of being able to say "to the best of knowledge, my builds don't need Java and please error out if it touches anything that touches Java".

That should already be possible today: Add `third_party/rules_java_stub/MODULE.bazel` with content `module(name = "rules_java")` to your repo and then either add `--override_module=%workspace%/third_party/rules_java_stub` to `.bazelrc` or add a `local_path_override` to `MODULE.bazel`. If `bazel vendor` doesn't skip overriden repos today, we could make it do so. Since the module file is empty, it won't register any toolchains and since there are no other files in the module, any references to `@rules_java` that were missed will result in a "thing not found" error.

**Comment by meteorcloudy** [comment]
@fmeum Ah, that's a good point and it's basically what I suggested at https://github.com/bazelbuild/bazel/issues/22684#issuecomment-2160961838. I can confirm `local_path_override` repo is already excluded from vendor.

Then `toolchain_exclude` may not be that necessary. I'll suggest the Kleaf team to try out this first.

**Comment by jacky8hyf** [comment]
This could work if we always use --override_module in the command-line. But unfortunately we can't put it as a default in .bazelrc or in MODULE.bazel local_path_override. This is because we also have stardoc() targets, [contained within a package](https://cs.android.com/android/kernel/superproject/+/common-android-mainline:build/kernel/kleaf/docs/BUILD.bazel). When building these targets, we allow Internet access because stardoc pulls a bunch of maven dependencies. So we only do that on developer's machines and regularly check-in the generated markdown files to the source tree. We do not build these stardoc() targets or even visit that package when in a contained environment (CI build machines without Internet access).

However, if we always override rules_java with a fake module, is there a way for us to offset the effect of --override_module when building stardoc? For example, can we make --override_module=rules_java= clear any previous values of --override_module for rules_java set in .bazelrc or in MODULE.bazel?

(on behalf of the Kleaf team)
